File: 5. Validations and transformations for backend engineers/code/typescript/example_6.ts
import { z } from 'zod';

export const PaginationSchema = z.object({
  // Zod coercing the incoming string "2" into int 2 (transform), 
  // THEN enforces gt/lt (validate)
  page: z.coerce.number().int().min(1).max(499),
  limit: z.coerce.number().int().min(1).max(9999)
});
export type Pagination = z.infer<typeof PaginationSchema>;

export function parsePagination(q: any): Pagination {
    return PaginationSchema.parse(q);
}

File: 5. Validations and transformations for backend engineers/code/rust/example_6.rs
use serde::Deserialize;
use validator::Validate;

#[derive(Debug, Deserialize, Validate)]
pub struct Pagination {
    // Serde automatically CASTS (transform) string from query param 
    // into int before we can VALIDATE the numbers.
    #[validate(range(min = 1, max = 499))]
    pub page: i32,
    #[validate(range(min = 1, max = 9999))]
    pub limit: i32,
}
